src/main/java/streamkv/operator/PendingOperationCheckpointer.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package streamkv.operator;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.TreeMap;
+
+import org.apache.flink.api.common.state.StateCheckpointer;
+import org.apache.flink.api.java.typeutils.runtime.ByteArrayInputView;
+import org.apache.flink.core.memory.DataOutputView;
+import org.apache.flink.core.memory.OutputViewDataOutputStreamWrapper;
+
+import streamkv.types.KVOperation;
+import streamkv.types.KVOperationTypeInfo.KVOpSerializer;
+
+public class PendingOperationCheckpointer<K, V> implements
+		StateCheckpointer<TreeMap<Long, List<KVOperation<K, V>>>, byte[]> {
+
+	private KVOpSerializer<K, V> opSerializer;
+
+	public PendingOperationCheckpointer(KVOpSerializer<K, V> opSerializer) {
+		this.opSerializer = opSerializer;
+	}
+
+	@Override
+	public byte[] snapshotState(TreeMap<Long, List<KVOperation<K, V>>> pending, long checkpointId,
+			long checkpointTimestamp) {
+		ByteArrayOutputStream bos = new ByteArrayOutputStream(pending.size() * 24);
```

Comment:
it's better to overallocate anyway. Now about reusing it, it might not be that a bad idea (and only update it when the state cardinality changes). We could run some tests to see